zrml/orderbook/src/lib.rs

```diff
@@ -164,181 +178,274 @@ mod pallet {
+
+        /// Charge the external fees from `taker` and return the adjusted maker fill.
+        fn charge_external_fees(
+            order_data: &OrderOf<T>,
+            base_asset: AssetOf<T>,
+            maker_fill: BalanceOf<T>,
+            taker: &AccountIdOf<T>,
+            taker_fill: BalanceOf<T>,
+        ) -> Result<BalanceOf<T>, DispatchError> {
+            let maker_asset_is_base = order_data.maker_asset == base_asset;
+            let base_asset_fill = if maker_asset_is_base {
+                taker_fill
+            } else {
+                debug_assert!(order_data.taker_asset == base_asset);
+                maker_fill
+            };
+            let fee_amount = T::ExternalFees::distribute(
+                order_data.market_id,
+                base_asset,
+                taker,
+                base_asset_fill,
+            );
+            if maker_asset_is_base {
+                // maker_fill is the amount what the maker wants to have (outcome asset from taker)
+                // do not charge fees from outcome assets, but rather from the base asset
+                Ok(maker_fill)
+            } else {
+                // accounting fees from the taker,
+                // who is responsible to pay the base asset minus fees to the maker.
+                Ok(maker_fill.checked_sub_res(&fee_amount)?)
+            }
+        }
+
+        fn do_fill_order(
+            order_id: OrderId,
+            taker: AccountIdOf<T>,
+            maker_partial_fill: Option<BalanceOf<T>>,
+        ) -> DispatchResult {
             let mut order_data = <Orders<T>>::get(order_id).ok_or(Error::<T>::OrderDoesNotExist)?;
             let market = T::MarketCommons::market(&order_data.market_id)?;
-            ensure!(market.scoring_rule == ScoringRule::Orderbook, Error::<T>::InvalidScoringRule);
+            debug_assert!(
+                market.scoring_rule == ScoringRule::Orderbook,
+                "The call to place_order already ensured the scoring rule order book."
+            );
             ensure!(market.status == MarketStatus::Active, Error::<T>::MarketIsNotActive);
             let base_asset = market.base_asset;
 
-            let makers_requested_total = match order_data.side {
-                OrderSide::Bid => order_data.outcome_asset_amount,
-                OrderSide::Ask => order_data.base_asset_amount,
-            };
-            let maker_fill = maker_partial_fill.unwrap_or(makers_requested_total);
-            ensure!(!maker_fill.is_zero(), Error::<T>::AmountIsZero);
-            ensure!(maker_fill <= makers_requested_total, Error::<T>::AmountTooHighForOrder);
+            let maker_fill = maker_partial_fill.unwrap_or(order_data.taker_amount);
+            ensure!(
+                maker_fill >= T::AssetManager::minimum_balance(order_data.taker_asset),
+                Error::<T>::BelowMinimumBalance
+            );
+            ensure!(maker_fill <= order_data.taker_amount, Error::<T>::AmountTooHighForOrder);
```

Comment:
Just as a side note, I find the naming of the maker and taker information very confusing, as it depends on the perspective that is being looked from. I would recommend sticking to an absolute view, for example `taker_fill` and `taker_amount` always refer to the token that the maker wants.

Reply:
This occurs in multiple places in this file.

Reply:
I don't really understand how to achieve this. I know it's at first hard to wrap the head around it. 

```
The `maker_partial_fill` is the partial amount of what the maker wants to get filled.
```

Reply:
I would just always use `maker` if it affects the maker token and always use `taker` if it affects the taker token.